File: stack/largest_rectangle_in_histogram/cpp/largest_rectangle_in_histogram.cpp
#include<iostream>
#include<vector>
#include<stack>

using std::cout;
using std::vector;

// Time O(n)
// Space O(n)

class Solution {
public:
    static int largestRectangleArea(const vector<int>& heights) {
        std::stack<std::pair<int, int>> stack;
        int answer = 0;

        // checks to area of the last biggest bar and return it's start
        auto checkStack = [&](int i) -> int {
            auto top = stack.top();
            int h = top.first;
            int w = i - top.second;
            int a = w * h;
            answer = std::max(answer, a);
            stack.pop();
            return top.second;
        };

        for(int i = 0; i < heights.size(); i++) {
            int start = i; // the start of the bar. Starts at i and gets changed to the prev 
                           // bar if the prev was bigger than current
            while(!stack.empty() && heights[i] < stack.top().first) {
                start = checkStack(i);
            }
            stack.push({ heights[i], start });
        }

        while(!stack.empty()) {
            checkStack(heights.size());
        }
        
        return answer;
    }
};

int main() {
    vector<int> heights = { 2, 1, 5, 6, 2, 3 };

    int answer = Solution::largestRectangleArea(heights);

    cout << answer << '\n';
}

File: stack/largest_rectangle_in_histogram/js/largest_rectangle_in_histogram.js
const heights = [2, 1, 5, 6, 2, 3];

// leetcode tells me this javascript is slow if anyone knows
// what I did wrong please make pr if you want to
const largestRectangleArea = (heights) => {
  let stack = [];
  let answer = 0;

  const checkStack = (i) => {
    let top = stack.pop();
    let h = top.height;
    let w = i - top.start;
    let a = h * w;
    answer = Math.max(answer, a);
    return top.start;
  }

  for(let i in heights) {
    let start = i;
    while(stack.length && heights[i] < stack.at(-1).height) {
      start = checkStack(i)
    }
    stack.push({ height: heights[i], start: start });
  }
  while(stack.length) {
    checkStack(heights.length);
  }

  return answer;
}

console.log(largestRectangleArea(heights));
